File: project/core/lib/common/src/main/java/vdi/core/health/DynamicDependency.java
package vdi.core.health;

import java.util.Map;
import java.util.function.Supplier;

public class DynamicDependency implements Dependency {
  private final String name;
  private final String protocol;
  private final String host;
  private final int port;
  private final Supplier<Map<String, Object>> extraSupplier;

  public DynamicDependency(
    String name,
    String protocol,
    String host,
    int port,
    Supplier<Map<String, Object>> extraSupplier
  ) {
    this.name = name;
    this.protocol = protocol;
    this.host = host;
    this.port = port;
    this.extraSupplier = extraSupplier;
  }

  @Override
  public String getName() {
    return this.name;
  }

  @Override
  public String getProtocol() {
    return this.protocol;
  }

  @Override
  public String getHost() {
    return this.host;
  }

  @Override
  public int getPort() {
    return this.port;
  }

  @Override
  public Map<String, Object> getExtraFields() {
    return this.extraSupplier.get();
  }

  @Override
  public Status checkStatus() {
    return Status.OK;
  }

  @Override
  public String toString() {
    return this.urlString();
  }
}

File: lib/commons/src/main/kotlin/vdi/lib/health/DynamicDependency.kt
package vdi.lib.health

open class DynamicDependency(
  override val name: String,
  override val protocol: String,
  override val host: String,
  override val port: UShort,
  val extraSupplier: () -> Map<String, Any>
) : Dependency {
  override val extra: Map<String, Any>
    get() = extraSupplier()

  override fun checkStatus() = Dependency.Status.Ok

  override fun toString() =
    if (protocol.isNotBlank())
      "$protocol://$host:$port"
    else
      "$host:$port"
}
